File: 2023/07_CamelCards/hand.py

# ======================================================================
# Camel Cards
#   Advent of Code 2023 Day 07 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         h a n d . p y
# ======================================================================
"Hand for the Advent of Code 2023 Day 07 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
from collections import Counter

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
CARD_ORDER = "AKQJT08765432"
CARD_STRENGTH = {
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "T": 10,
    "J": 11,
    "Q": 12,
    "K": 13,
    "A": 14,
}
CARD_STRENGTH_PART_TWO = {
    "2": 2,
    "3": 3,
    "4": 4,
    "5": 5,
    "6": 6,
    "7": 7,
    "8": 8,
    "9": 9,
    "T": 10,
    "J": 1,
    "Q": 12,
    "K": 13,
    "A": 14,
}
FIVE_OF_A_KIND = 7
FOUR_OF_A_KIND = 6
FULL_HOUSE = 5
THREE_OF_A_KIND = 4
TWO_PAIR = 3
ONE_PAIR = 2
HIGH_CARD = 1

# ======================================================================
#                                                                   Hand
# ======================================================================


class Hand(object):   # pylint: disable=R0902, R0903, R0205
    "Object for Camel Cards"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.cards = ""
        self.bid = 0
        self.type = 0
        self.key = 0

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self._process_text(text)

    def _process_text(self, text):
        "Assign values from text"

        assert text is not None and len(text) > 0

        # 1. Split into cards and bid
        parts = text.split()
        assert len(parts) == 2
        assert len(parts[0]) == 5

        # 2. Save them
        self.cards = parts[0]
        self.bid = int(parts[1])

        # 3. Determine the type and key
        if self.part2:
            self.type = self.determine_type_part_two(self.cards)
        else:
            self.type = self.determine_type(self.cards)
        self.key = self.determine_key()

    def determine_type(self, cards):  # pylint: disable=R0911
        "Return the type for this hand"

        # 1. If we already know it, return the stored value
        if self.type > 0:
            return self.type

        # 2. Count the cards
        knt = Counter(cards)
        oknt = knt.most_common()

        # 3. Is it five of a kind?
        if oknt[0][1] == 5:
            return FIVE_OF_A_KIND

        # 4. Is it four of a kind?
        if oknt[0][1] == 4:
            return FOUR_OF_A_KIND

        # 5. Is it three of a kind or full house?
        if oknt[0][1] == 3:
            if oknt[1][1] == 2:
                return FULL_HOUSE
            return THREE_OF_A_KIND

        # 6. Do we have one or two pairs?
        if oknt[0][1] == 2:
            if oknt[1][1] == 2:
                return TWO_PAIR
            return ONE_PAIR

        # 7. Which leaves only high card
        return HIGH_CARD

    def determine_type_part_two(self, cards):
        "Return the type for this hand part two hand"

        # 1. If we already know it, return the stored value
        if self.type > 0:
            return self.type

        # 2. If the hand has no jokers, evaluate it as normal
        if "J" not in cards:
            return self.determine_type(cards)

        # 3. Assume the least of the types
        result = 0

        # 4. Loop for all of the cards
        for card in cards:

            # 5. Set all the jokers to this value
            jcards = cards.replace("J", card)

            # 6. Determine the type with this hand of cards
            jtype = self.determine_type(jcards)

            # 7. Keep the highest type
            if jtype > result:
                result = jtype

        # 8. Return the highest type
        return result

    def determine_key(self):
        "Returns the sorting key of this hand"

        # 1. If we know the key, return it
        if self.key > 0:
            return self.key

        # 2. Get the card values
        if self.part2:
            values = [CARD_STRENGTH_PART_TWO[x] for x in self.cards]
        else:
            values = [CARD_STRENGTH[x] for x in self.cards]

        # 3. Combine the card values
        result = self.determine_type(self.cards)
        for value in values:
            result = result * 100 + value

        # 4. Return the computed key
        return result

    def winnings(self, rank):
        "Return the winnings for this hand"
        return rank * self.bid


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                          h a n d . p y                         end
# ======================================================================

File: 2023/07_CamelCards/hand.go
// ======================================================================
// Camel Cards
//   Advent of Code 2023 Day 07 -- Eric Wastl -- https://adventofcode.com
//
// go implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                         h a n d . g o
// ======================================================================
// Hand for the Advent of Code 2023 Day 07 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
	"strconv"
	"strings"
)

// ----------------------------------------------------------------------
//
//	constants
//
// ----------------------------------------------------------------------
const (
	handFive   = "Five of a kind"
	handFour   = "Four of a kind"
	handFull   = "Full house"
	handThree  = "Three of a kind"
	handTwo    = "Two pair"
	handOne    = "One pair"
	handHigh   = "High card"
	handNone   = "No cards"
	cardOrder1 = "AKQJT98765432"
	cardOrder2 = "AKQT98765432J"
	joker      = "J"
)

var (
	handType = map[string]int{
		handFive:  6,
		handFour:  5,
		handFull:  4,
		handThree: 3,
		handTwo:   2,
		handOne:   1,
		handHigh:  0,
		handNone:  0,
	}
)

// ======================================================================
//                                                                   Hand
// ======================================================================

// Hand ... Object for Camel Cards
type Hand struct {
	Part2 bool
	Text  string
	Cards string
	Bid   int
	Type  int
}

// NewHand ... Create a Hand object
func NewHand(part2 bool, text string) (*Hand, error) {
	// Hand Object for Camel Cards

	// 1. Set the initial values
	p := new(Hand)
	p.Part2 = part2
	p.Text = strings.TrimSpace(text)

	// 2. Set the Cards and Bid
	if len(p.Text) > 0 {
		fields := strings.Fields(p.Text)
		if len(fields) != 2 {
			return nil, fmt.Errorf("hand '%s' should have two values not %d", p.Text, len(fields))
		}
		p.Cards = fields[0]
		bid, err := strconv.Atoi(fields[1])
		if err != nil {
			return nil, fmt.Errorf("invalid bid value '%s' for hand '%s", fields[2], p.Text)
		}
		p.Bid = bid
	}

	// 3. Determine the type of the hand
	p.Type = p.HandType()

	// 4. Return success
	return p, nil
}

// HandType ... Returns the numerical hand type
func (p *Hand) HandType() int {

	// 1. Determine the type of hand
	hType := p.HandType1()
	if p.Part2 {
		hType = p.HandType2()
	}

	// 2. Return the numeric type
	return handType[hType]
}

// HandType1 ... Returns the hand type (No Jokers)
func (p *Hand) HandType1() string {

	// 1. Really should have some cards
	if len(p.Cards) == 0 {
		return handNone
	}

	// 2. Get the count of the cards and the counts of the counts
	countCards := p.CountCards()
	countCounts := CountCounts(countCards)

	// 3. Do we have Five of a kind?
	if countCounts[5] > 0 {
		return handFive
	}

	// 4. Do we have Four of a kind?
	if countCounts[4] > 0 {
		return handFour
	}

	// 5. Do we have a Full house?
	if countCounts[3] > 0 && countCounts[2] > 0 {
		return handFull
	}

	// 6. Do we have Three of a kind?
	if countCounts[3] > 0 {
		return handThree
	}

	// 7. Do we have Two pair?
	if countCounts[2] > 1 {
		return handTwo
	}

	// 8. Do we have One pair?
	if countCounts[2] > 0 {
		return handOne
	}

	// 9. Else it is High card
	return handHigh
}

// HandType2 ... Returns the hand type (With Jokers)
func (p *Hand) HandType2() string {

	// 1. If no Jokers, just use the regular handtype
	if !strings.Contains(p.Cards, joker) {
		return p.HandType1()
	}

	// 2. Start with a low hand
	result := handHigh
	originalCards := p.Cards

	// 3. Loop for the possible replacements for the Joker
	for _, card := range cardOrder1 {
		replacement := string(card)
		if replacement == joker {
			continue
		}

		// 4. Replace the jokers
		p.Cards = strings.ReplaceAll(originalCards, joker, replacement)

		// 5. Get the type of this modified hand
		jokerType := p.HandType1()

		// 6. If this better than what we had, save it
		if handType[jokerType] > handType[result] {
			result = jokerType
		}
	}

	// 7. Return the best hand type using Jokers
	p.Cards = originalCards
	return result
}

// CountCards ... Returns the number of each card
func (p *Hand) CountCards() map[string]int {

	// 1. Start a map for the cards
	counts := make(map[string]int, len(cardOrder1))
	for indx := 0; indx < len(cardOrder1); indx = indx + 1 {
		card := string(cardOrder1[indx])
		counts[card] = 0
	}

	// 2. Loop for all the cards in the hand
	for indx := 0; indx < len(p.Cards); indx = indx + 1 {
		card := string(p.Cards[indx])

		// 3. Increment the count of the card
		counts[card] = counts[card] + 1
	}

	// 4. Return the counts of each of the cards
	return counts
}

// Less ... Returns true of p's hand is less than o's
func (p *Hand) Less(o *Hand) bool {

	// 1. Compare the rank of the hand types
	if p.Type < o.Type {
		return true
	}
	if p.Type > o.Type {
		return false
	}

	// 2. Compare the hands card by card
	for indx := 0; indx < len(p.Cards); indx = indx + 1 {

		// 3. If the cards are not equal, return true or false
		if p.CardLess(o, indx) {
			return true
		}
		if o.CardLess(p, indx) {
			return false
		}
	}

	// 3. Should never reach error
	fmt.Printf("unexpected equality between '%s' %d and '%s' %d\n", p.Cards, p.Type, o.Cards, o.Type)
	return true
}

// CardLess ... Return true if the indx'th card of p is less than the one from o
func (p *Hand) CardLess(o *Hand, indx int) bool {

	// 1. Return true if bad indx
	if indx < 0 || indx >= len(p.Cards) || indx >= len(o.Cards) {
		fmt.Printf("unexpected index %d for '%s' and '%s'\n", indx, p.Cards, o.Cards)
		return true
	}

	// 2. What are we using to rank the cards?
	order := cardOrder1
	if p.Part2 {
		order = cardOrder2
	}

	// 3. Get the order of the cards
	oCard := string(o.Cards[indx])
	oOrder := strings.Index(order, oCard)
	if oOrder == -1 {
		fmt.Printf("unexpected card '%s' at index %d of '%s', not in '%s'", oCard, indx, o.Cards, order)
		return true
	}
	pCard := string(p.Cards[indx])
	pOrder := strings.Index(order, pCard)
	if pOrder == -1 {
		fmt.Printf("unexpected card '%s' at index %d of '%s', not in '%s'", pCard, indx, p.Cards, order)
		return true
	}

	// 4. Return difference pCard to oCard (note order string is reversed)
	return pOrder > oOrder
}

// CountCounts ... Return the count of counts
func CountCounts(countCards map[string]int) map[int]int {

	// 1. Start all zeroes
	counts := make(map[int]int, 6)
	for indx := 0; indx < 6; indx = indx + 1 {
		counts[indx] = 0
	}

	// 2. Loop for the count of each card
	for indx := 0; indx < len(cardOrder1); indx = indx + 1 {
		card := string(cardOrder1[indx])
		count := countCards[card]

		// 3. Increment the count of the counts
		counts[count] = counts[count] + 1
	}

	// 4. Return the count of the counts
	return counts
}

// ======================================================================
// end                          h a n d . g o                         end
// ======================================================================
